src/main/java/com/iota/iri/utils/collections/impl/BoundedSetWrapper.java

```diff
@@ -0,0 +1,172 @@
+package com.iota.iri.utils.collections.impl;
+
+import com.iota.iri.utils.collections.interfaces.BoundedSet;
+
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+
+/**
+ * A wrapper for a set that allows up to {@code maxsize} elements.
+ * Old elements are removed once it is full and new elements are added.
+ * The way old elements are chosen depends on the underlying set implementation.
+ *
+ *
+ * @param <E> the type parameter
+ */
+public class BoundedSetWrapper<E> implements BoundedSet<E>{
+    private final int maxSize;
+    private final Set<E> delegate;
+
+    /**
+     * Wraps the given set {@code c}
+     *
+     * @param c the set which you delegate the actions to
+     * @param maxSize the max size
+     * @throws NullPointerException if the {@code c} is null
+     * @throws IllegalArgumentException if {@code c} is larger than {@code maxSize}
+     */
+    public BoundedSetWrapper(Set<E> c, int maxSize) {
+        Objects.requireNonNull(c, "trying to wrap a null set");
+        requireCollectionIsNotAboveMaxSize(c, maxSize);
+        this.maxSize = maxSize;
+        this.delegate = c;
+    }
+
+    @Override
+    public int getMaxSize() {
+        return maxSize;
+    }
+
+    @Override
+    public int size() {
+       return delegate.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return delegate.contains(o);
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return delegate.iterator();
+    }
+
+    @Override
+    public void forEach(Consumer<? super E> action) {
+        delegate.forEach(action);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return delegate.toArray();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        return delegate.toArray(a);
+    }
+
+    @Override
+    public boolean add(E e) {
+        if (isFull()) {
+            Iterator<E> iterator = delegate.iterator();
+            iterator.next();
+            iterator.remove();
+        }
+
+        return delegate.add(e);
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return delegate.remove(o);
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        return delegate.containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends E> c) {
+        requireCollectionIsNotAboveMaxSize(c, getMaxSize());
+        if (canCollectionBeFullyAdded(c)) {
+            return delegate.addAll(c);
+        }
+        else {
+            int itemsToDelete = delegate.size() + c.size() - getMaxSize();
```

Comment:
so I don't forget - please open an issue (not blocking the PR):
there may be duplicate entries in the sets `delegate` and `c` - in that case you may want to remove less items.
for example:
```
delegate = {a, b, c, d}
c = {c, d, e}
maxSize = 5
itemsToDelete = 4 +3 - 5 => 2
result: {c, d, e}
```

you could argue that it's too costly to check, but then there should be a comment explaining that heuristic.

Reply:
Opened https://github.com/iotaledger/iri/issues/887